File: answers/leetcode/Swap Nodes in Pairs/Swap Nodes in Pairs.java
/**
 * Definition for singly-linked list.
 * public class ListNode {
 *     int val;
 *     ListNode next;
 *     ListNode(int x) { val = x; }
 * }
 */
class Solution {
    public ListNode swapPairs(ListNode head) {
        ListNode dummy = new ListNode(-1);
        dummy.next = head;
        ListNode current = dummy;
        while (current != null && current.next != null && current.next.next != null) {
            ListNode temp = current.next;
            current.next = current.next.next;
            temp.next = temp.next.next;
            current.next.next = temp;
            current = current.next.next;
        }
        return dummy.next;
    }
}

File: answers/leetcode/Swap Nodes in Pairs/Swap Nodes in Pairs.cpp
//@result 55 / 55 test cases passed. Status: Accepted Runtime: 4 ms Submitted: 0 minutes ago You are here! Your runtime beats 2.50% of cpp submissions.

/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode(int x) : val(x), next(NULL) {}
 * };
 */
class Solution {
public:
    ListNode *swapPairs(ListNode *head) {
        // Start typing your C/C++ solution below
        // DO NOT write int main() function
		if (NULL == head || NULL == head->next)
		{
			return head;
		}
		ListNode *tail = swapPairs(head->next->next);
		ListNode *new_head = head->next;
		new_head->next = head;
		new_head->next->next = tail;
		return new_head;
    }
};
